doc/cli_wrapper/shd-cli_wrapper_hld.md

```diff
@@ -0,0 +1,213 @@
+
+- [3. Open questions](#3-open-questions)
+
+- [4. Reference](#4-reference)
+
+<!-- markdown-toc end -->
+
+## Revision
+
+| Rev  |    Date    |                Author                 | Change Description |
+| :--: | :--------: | :-----------------------------------: | ------------------ |
+| 0.1  | 08/12/2022 |  Justin Lu                            | 1. Initial version    |
+| 0.2  | 09/03/2022 |  Justin Lu                            | 1. Add watchdog design <br> 2. Modify architecture picture |
+| 0.3  | 09/22/2022 |  Antonio Ho                           | 1. Add enable/disable design |
+| 0.4  | 01/16/2023 |  Antonio Ho                           | 1. Add performance improvements |
+| 0.5  | 05/03/2023 |  Antonio Ho                           | 1. Modify performance improvements <br> 2. Add TODO list |
+
+# 1. High Level Design Document
+
+## 1.1 About this Manual
+
+This document provides the high-level design for CLI_wrapper in SONiC
+
+## 1.2 Architecture
+
+SONiC uses click package as its CLI engine, although click package provides developers a convenient way to establish whole CLI architecture, it also has some disadvantage. The problem we want to address here is about the bad performance of CLI execution time. 
+
+Everytime user executes a command, CLI engine will load main source file (ex. config/main.py), then starts to execute the command. In this process, all libraries always have to be imported again. It costs about 80% execution time on this process, so CLI_wrapper is designed to reduce the time.
+
+The figure below depicts the architecture of CLI_wrapper. 
+
+![Architecture](images/architecture_1.jpg)
+
+* CLI_wrapper builder
+
+  In order to integrate CLI_wrapper with current CLI architecture, we develop CLI_wrapper builder to generate CLI_wrapper for command source file (ex. main.py). When sonic-utilities debian is packaged, it calls CLI_wrapper builder to generate wrapped files which contains only function interface and minimal libraries. It replaces function body with the calling of CLI_wrapper client to send a encoded command to CLI_wrapper server and acquires the result. Since the CLI_wrapper only contains minimal libraries, CLI engine doesn't need a lot of time to execute.
+
+![Architecture](images/architecture_2.jpg)
+
+* CLI_wrapper server
+
+  CLI_wrapper server is responsible for the execution of click command. It starts when system boots up, and loads all needed libraries to server. Then setup a socket to wait client's request. Once it accepts a request, it spawns a CLI_worker process to deal with this request. 
+  
+  In order to avoid that some command may stuck in CLI worker and cause resource leakage, a watchdog thread is designed to watch the execution time of CLI worker, once the execution time exceeds maximum execution time, the CLI worker will be gracefully terminated. But if the CLI worker is still alive after one minute, a kill signal will be sent to CLI worker to tear down it. 

```

Comment:
There are config/show commands that can go beyond 1 min, hope you have a configuration to increase the timeout if required.

Reply:
When the show output has more than 1 page, how are you handling the pagination? can we make sure the current behavior is not disturbed with this new design?